Code Refactor

aquatiko opened the pull request:
Upon completion of the prototyping phase, this PR aims to refactor the experimental code-base.

aquatiko committed:
Code Refactor

Encapsulate logically seperate tasks into sperate modules.

aquatiko committed:
Add docstrings

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/AnalyserActivity.java:
```diff
@@ -0,0 +1,125 @@
+    analyseResultField.append(
+        "\nPort bound:" + CameraActivity.serverHandler.getServerSocketPort() + "\n");
+
+    CameraFragment.Companion.getFilePath();
```
Why do you call this twice?

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/AnalyserActivity.java:
```diff
@@ -0,0 +1,125 @@
+    filePath.setText(CameraFragment.Companion.getFilePath());
+
+    filePickerBtn.setOnClickListener(
+        new View.OnClickListener() {
```
Are you able to change to using lambdas for these?  There is a lot of low information content boilerplate code here.

I haven't personally done it before, but for example: https://mayojava.github.io/android/java/using-java8-lambda-expressions-in-android/

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/AnalyserActivity.java:
```diff
@@ -0,0 +1,125 @@
+          public void onClick(View view) {
+            try {
+              analyze();
+            } catch (Exception e) {
```
Are you expecting any particular exceptions here?

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/AnalyserActivity.java:
```diff
@@ -0,0 +1,125 @@
+    if (requestCode == FILE_PICKER_REQUEST_CODE && resultCode == RESULT_OK) {
+      String path = data.getData().toString();
+      fileUri = data.getData();
+      filePath.setText(path);
```
filePath.setText(fileUri.toString()); ?  Then you can remove the first line and only call getData once.

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/AnalyserActivity.java:
```diff
@@ -0,0 +1,125 @@
+  /** Triggers subsequent lifecycle events of lag calculations. */
+  @RequiresApi(api = Build.VERSION_CODES.P)
+  private void analyze() {
+    fileUri = Uri.parse(filePath.getText().toString());
```
Why is fileUri a member variable if we overwrite it every time we call analyze()?

patrickmeiring replied:
Agree. In this version of the code, I'd probably vote to remove fileUri as a class-level variable and keep only one copy of state (namely in the textbox). Down the track, when we integrate capture and analysis, and delete the file picker, we can switch back to using a fileUri if need be.

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/AnalyserActivity.java:
```diff
@@ -0,0 +1,125 @@
+  @RequiresApi(api = Build.VERSION_CODES.P)
+  private void analyze() {
+    fileUri = Uri.parse(filePath.getText().toString());
+    videoProcessor.setVideoReader(getApplicationContext(), fileUri);
```
A setVideoReader function should take a VideoReader argument, perhaps createVideoReader is a beter name?

aquatiko replied:
done

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/AnalyserActivity.java:
```diff
@@ -0,0 +1,125 @@
+    fileUri = Uri.parse(filePath.getText().toString());
+    videoProcessor.setVideoReader(getApplicationContext(), fileUri);
+
+    CameraActivity.serverHandler.downloadServerTimeStamps();
```
Can we make ServerHandler a singleton rather than piggybacking off of the CameraActivity singleton?

It doesn't really make sense when reading the code why this needs to reference CameraActivity.  It should at least be commented, but ideally this file shouldn't need to reference CameraActivity at all.  I believe Activities in Android should be independent and only communicate by Intents.

patrickmeiring replied:
I personally wouldn't bother with the Singleton if the intent is to merge CameraActivity & AnalyserActivity.

While I don't like the way the code is now (activities should only communicate with intents), the issue existed in the old code (MainActivity was using the sockets on CameraActivity directly) and this would all become local to the activity when CameraActivity & AnalyserActivity are merged. Maybe put in a TODO though.

aquatiko replied:
done

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/AnalyserActivity.java:
```diff
@@ -0,0 +1,125 @@
+
+    CameraActivity.serverHandler.downloadServerTimeStamps();
+
+    videoProcessor.doOcrOnVideo(lagCalculator);
```
Just doOcr ?  Also should it be doOCR? not sure what the Java style here is.

aquatiko replied:
done

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/AnalyserActivity.java:
```diff
@@ -0,0 +1,125 @@
+
+    CameraActivity.serverHandler.downloadServerTimeStamps();
+
+    videoProcessor.doOcrOnVideo(lagCalculator);
```
lagCalculator already knows about the videoProcessor (passed in the constructor), why do you need to make them mutually dependant?  try to make your dependencies go only one way (dependency graph should be acyclic)

patrickmeiring replied:
Similar comments on the class structure, but I think we should go one step further. 
Either:
- Encapsulate videoProcessor inside lagCalculator (or vica-versa), or 
- Make videoProcessor and lagCalculator independent objects (that don't reference eachother) and let the AnalyserActivity pass results from one to the other.

At the moment, it looks like videoProcessor is used from both inside lagCalculator and in AnalyserActivity. This feels like AnalyserActivity is forming part of the implementation of lagCalculator (i.e. lagCalculator fails to abstract AnalyserActivity from its implementation details; we must read both classes together to understand how either work).

If AnalyserActivity must delegate videoProcessor and lagCalculator to communicate, prefer to use an interface that makes clear what communication that is occurring and decouples LagCalculator from VideoProcessor.

Overall, the flow that seems most logical to me right now is:

Timestamps inputTimestamps = serverHandler.getInputTimestamps();
OcrResults videoTimestamps = VideoProcessor.getVideoTimestamps(video);
LagResults inputLags = LagCalculator.computeLags(inputTimestamps, videoTimestamps);

Where Timestamps/OcrResults/Lags are only container classes that hold data, and VideoProcessor and LagCalculator are static classes (have no state).

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/AnalyserActivity.java:
```diff
@@ -0,0 +1,125 @@
+    fileUri = Uri.parse(filePath.getText().toString());
+    videoProcessor.setVideoReader(getApplicationContext(), fileUri);
+
+    CameraActivity.serverHandler.downloadServerTimeStamps();
```
Would probably be better if this method returned the timestamps and you passed them as a parameter to where they are needed.  When you do it this way you are storing state in the object and it may not be obvious when something doesn't work that it is because this function wasn't called.

patrickmeiring replied:
Agree. Prefer to make the data flow obvious through method results/parameters, not flow in a hidden way.

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/AnalyserActivity.java:
```diff
@@ -0,0 +1,125 @@
+    @Override
+    protected Void doInBackground(LagCalculator... lagCalculators) {
+      while (true) {
+        if (resultPublishedCount < lagCalculators[0].lagResult.size()) {
```
You are only using the 0th element:
1. How do you know the list isn't empty?
2. Why aren't you using the rest?

If it is only a single object, please try to pass a single object rather than an array of size 1, and if not possible, please document why it is not possible, and the answers to the above two points.

mwarton replied:
Having a loop counter as a class member variable is a pretty far from normal coding construct.  Please make it local or document clearly why it needs to be this way.

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/AnalyserActivity.java:
```diff
@@ -0,0 +1,125 @@
+      while (true) {
+        if (resultPublishedCount < lagCalculators[0].lagResult.size()) {
+          analyseResultField.append(
+              "Char = "
```
use String.format instead.

mwarton reviewed (commented):
I've only looked at the first file, probably more to come tomorrow

patrickmeiring commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/AnalyserActivity.java:
```diff
@@ -0,0 +1,125 @@
+    analyzeBtn = findViewById(R.id.analyze);
+    analyseResultField = findViewById(R.id.analyseResultView);
+
+    analyseResultField.setMovementMethod(new ScrollingMovementMethod());
```
I'm somewhat partial to doing this all in XML if possible, and leaving the code behind to business logic as much as possible. How about wrapping your TextView in a ScrollView and removing this?
Various options are discussed here: https://stackoverflow.com/questions/1748977/making-textview-scrollable-on-android

patrickmeiring commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/AnalyserActivity.java:
```diff
@@ -0,0 +1,125 @@
+
+    @Override
+    protected Void doInBackground(LagCalculator... lagCalculators) {
+      while (true) {
```
Why not make this a for loop? That seems simpler and would make loop termination clearer.
for (int resultPublishedCount = 0; resultPublishedCount < ... .size(); resultPublishedCount++) {
  ...
}
You could probably then also rename resultPublishedCount to 'i'.

patrickmeiring reviewed (commented):
Good work, just taking a break from reviewing as well as it is getting late. 

To help us, could you also provide a statement of what this refactoring is (and is not?) intending to do?

aquatiko committed:
remove file picker feature

aquatiko commented:
Working on merging `AnalyserActivity` and `CameraActivity`. 
That will fix some of the code-reviews by simplifying the logic and rest will be fixing along with it too.

patrickmeiring commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/LagCalculator.java:
```diff
@@ -0,0 +1,50 @@
+    // Finds the video frame with the running serverSequence text.
+    for (int j = 1; j < serverTimestamp.size(); j++) {
+      for (int i = 0; i < resultsOCR.size(); i++) {
+        if (serverSequence.equals(resultsOCR.get(i))) {
```
It seems like this may not behave as expected if two characters were to appear on the screen at the same time. Thoughts about perhaps doing prefix matching?

aquatiko replied:
done

patrickmeiring commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/ServerHandler.java:
```diff
@@ -0,0 +1,111 @@
+/** Handles every interaction of app with the server. */
+public class ServerHandler {
+
+  private Integer serverSocketPort;
```
Prefer 'int' over Integer here, as there is no need for this to be nullable. (Also update corresponding getters and setters.)

aquatiko replied:
done

patrickmeiring commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/ServerHandler.java:
```diff
@@ -0,0 +1,111 @@
+  private Integer serverSocketPort;
+  private final String SERVER_IP = "127.0.0.1";
+  private ArrayList<Long> serverTimeStamps = new ArrayList<>();
+  private Long serverStartTimeStamp;
```
Ditto, prefer the unboxed type here.

aquatiko replied:
done

patrickmeiring commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/ServerHandler.java:
```diff
@@ -0,0 +1,111 @@
+  private Socket socket;
+  public BufferedReader inputReader;
+  public PrintWriter outputWriter;
+  private Long hostSyncTimeStamp;
```
Ditto (long instead of Long)

aquatiko replied:
done

patrickmeiring commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/ServerHandler.java:
```diff
@@ -0,0 +1,111 @@
+  private Long serverStartTimeStamp;
+  private ConnectionThread connectionThread;
+  private Socket socket;
+  public BufferedReader inputReader;
```
Why is this field public? If you need to expose the socket reader/writer, prefer to write a getter that returns them, and let external users do: serverHandler.getInputReader().read(...). 
I don't think external users of this class should be able to set this field.

aquatiko replied:
done

patrickmeiring commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/ServerHandler.java:
```diff
@@ -0,0 +1,111 @@
+
+  public void startConnection() {
+    connectionThread = new ConnectionThread();
+    connectionThread.execute();
```
It seems like we never wait for this task to complete. How does the consumer of this class know when it is safe to call downloadServerTimeStamps()? Do we need a join or await here?

patrickmeiring commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/ServerHandler.java:
```diff
@@ -0,0 +1,111 @@
+  public PrintWriter outputWriter;
+  private Long hostSyncTimeStamp;
+
+  public Long getHostSyncTimeStamp() {
```
Timestamp is one word, so this should be "Timestamp" not "TimeStamp"
https://google.github.io/styleguide/javaguide.html#s5.3-camel-case

(please do find and replace through this class, I see many occurances.)

aquatiko replied:
done

patrickmeiring commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/VideoProcessor.java:
```diff
@@ -0,0 +1,119 @@
+ */
+public class VideoProcessor {
+
+  private static final Integer SAFE_FRAMES = 10;
```
Please to use 'int' over "Integer'; there is no need for this to be nullable (ditto for occurances below)

aquatiko replied:
done

patrickmeiring commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/VideoProcessor.java:
```diff
@@ -0,0 +1,119 @@
+
+  public void doOcrOnVideo(LagCalculator lagCalculator) {
+    AnalyseVideo ocrOnVideoTask = new AnalyseVideo();
+    ocrOnVideoTask.execute(lagCalculator);
```
How does the client know this task has completed and the results are available? You're not awaiting or returning a handle to the AsyncTask that is returned.

patrickmeiring commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/VideoProcessor.java:
```diff
@@ -0,0 +1,119 @@
+
+  private Long recordStartTime = CameraFragment.Companion.getRecordingStartMillis();
+  private Long frameDuration = 1000L / CameraFragment.Companion.getFpsRecording();
+  private Integer framesProcessed = 0;
```
Ditto x3

aquatiko replied:
done

patrickmeiring reviewed (commented):
Added some more comments. Some of the comments may be a little bit superfluous as they relate to preexisting issues (so we can defer addressing them until later).

aquatiko committed:
Merge branch 'master' of https://github.com/googleinterns/screen-latency-testing into refactoring

aquatiko committed:
remove AnalyserActivity

As there is no need of AnalyserActivity UI now, migrated the core logic
to CameraActivity which can be started on completion of camera_fragment

aquatiko committed:
prefix search, names fix, lambdas, type migration

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/CameraActivity.kt:
```diff
@@ -16,41 +16,32 @@
+        val laptopPort = bundle?.getString("port")?.toInt() ?: 0
         if (laptopPort != 0) {
-            connectionThread.start()
+            serverHandler.serverSocketPort = laptopPort
```
It would be better to pass this in the constructor for serverHandler.  Is there any reason you are constructing it before onCreate?

aquatiko replied:
No specific reason. Will fix this.

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/CameraActivity.kt:
```diff
@@ -63,6 +54,19 @@ class CameraActivity : AppCompatActivity() {
         }, IMMERSIVE_FLAG_TIMEOUT)
     }
 
+    internal fun analyse(fileUri: Uri){
```
probably better to use the american spelling: analyze

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/CameraActivity.kt:
```diff
@@ -63,6 +54,19 @@ class CameraActivity : AppCompatActivity() {
+            videoProcessor.createVideoReader(applicationContext, fileUri)
+        };
+
+        serverHandler.downloadServerTimeStamps()
```
can you return and pass the timestamps as an argument?

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/CameraActivity.kt:
```diff
@@ -63,6 +54,19 @@ class CameraActivity : AppCompatActivity() {
+        val videoProcessor = VideoProcessor()
+        val lagCalculator = LagCalculator(videoProcessor, serverHandler)
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
```
What if this is not true?  Shouldn't an error be raised or something?

Does continuing on in this scenario make sense?

Can we just mark the App as requiring this build version and remove this (and any other similar) check?

aquatiko replied:
Using `@RequiresApi(Build.VERSION_CODES.P)` will make sure to add the same to app API level requirment.
Thanks for pointing. Will add this in subsequent commit.

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/LagCalculator.java:
```diff
@@ -0,0 +1,51 @@
+ */
+public class LagCalculator {
+  private VideoProcessor videoData;
+  private ServerHandler serverData;
```
maybe call it serverHandler rather than serverData, it is more than just some data.

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/LagCalculator.java:
```diff
@@ -0,0 +1,51 @@
+ * character seen timestamp from recorded video.
+ */
+public class LagCalculator {
+  private VideoProcessor videoData;
```
maybe call it videoProcessor

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/LagCalculator.java:
```diff
@@ -0,0 +1,51 @@
+public class LagCalculator {
+  private VideoProcessor videoData;
+  private ServerHandler serverData;
+  public ArrayList<Long> lagResult = new ArrayList<Long>();
```
array names should be plural: lagResults

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/LagCalculator.java:
```diff
@@ -0,0 +1,51 @@
+    this.serverData = serverData;
+  }
+
+  public ArrayList<Long> calculateLag() {
```
This should be a static method that takes all the data it needs as parameters.  There is no need for any of the class variables here, they can only introduce subtle bugs later.

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/LagCalculator.java:
```diff
@@ -0,0 +1,51 @@
+      for (int i = 0; i < resultsOCR.size(); i++) {
+        if (resultsOCR.get(i).startsWith(serverSequence)) {
+          long lag = videoFrameTimestamp.get(i) - serverTimestamp.get(j) + serverHostSyncOffset;
+          lagResult.add(lag);
```
Since this is a member variable, if you call this function twice you will get the results from both calls returned in the second call.  Please make lagResult a local variable only.

aquatiko replied:
fixed

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/LagCalculator.java:
```diff
@@ -0,0 +1,51 @@
+          break;
+        }
+      }
+      serverSequence += "m";
```
Please leave a todo to let the server inform the app what text to be expecting each time rather than hardcoding it.  (doesn't need to fix in this PR)

aquatiko replied:
Already added!

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/LagCalculator.java:
```diff
@@ -0,0 +1,51 @@
+
+    ArrayList<Long> serverTimestamp = serverData.getServerTimestamps();
+    ArrayList<Long> videoFrameTimestamp = videoData.getVideoFrameTimestamp();
+    ArrayList<String> resultsOCR = videoData.getResultsOCR();
```
How is multiple text detected by OCR in a single frame handled?  We need to be resilient to this in the final app.  Doesn't need to be in this PR, but please start thinking about it.

aquatiko committed:
synchronous execution and making data flow obvious

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/CameraActivity.kt:
```diff
@@ -16,41 +16,34 @@
-            e.printStackTrace()
-        }
-    })
+    internal lateinit var serverHandler: ServerHandler
```
can this be private?

aquatiko replied:
I'm keeping this non-private becasue `camera_fragment`(viz. an associated fragment to this CameraActivity) needs to use `serverHandler` and making it private will not allow the fragment to access it's own activity object.

aquatiko replied:
Ref [L#290](https://github.com/googleinterns/screen-latency-testing/blob/e4acf3f854f780729be639a644f29c2d592f1aa2/Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/fragments/CameraFragment.kt#L290)  [L#332](https://github.com/googleinterns/screen-latency-testing/blob/e4acf3f854f780729be639a644f29c2d592f1aa2/Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/fragments/CameraFragment.kt#L332)

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/LagCalculator.java:
```diff
@@ -0,0 +1,81 @@
+
+  private class LagCalculatorThread extends AsyncTask<LagCalculatorPrams, Void, ArrayList<Long>> {
+    @Override
+    protected ArrayList<Long> doInBackground(LagCalculatorPrams... params) {
```
Why are you doing this in the background?  This step is just a few hundred string comparisons right?

aquatiko replied:
Simplified in a non-async part.

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/LagCalculator.java:
```diff
@@ -0,0 +1,81 @@
+      return lagCalculatorThread
+          .execute(
+              new LagCalculatorPrams(serverTimestamps, videoFrameTimestamp, resultsOcr, syncOffset))
+          .get();
```
calling get blocks the main thread waiting for the result, ie it is exactly the same as calling this directly without putting it in a separate task, as mentioned in the other comment, I don't think anything in this class needs to be async.

aquatiko replied:
Simplified in a non-async part.

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/LagCalculator.java:
```diff
@@ -0,0 +1,81 @@
+
+    public LagCalculatorPrams(
+        ArrayList<Long> serverTimestamps,
+        ArrayList<Long> videoFrameTimestamps,
```
I think you should create a data structure that holds both the timestamp and the text, and have a single array.  This will reduce the number of params and should also make the code simpler. (no need for indexes in your calculation loop)

aquatiko replied:
This is simplified without an async-task in the following commit.

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/ServerHandler.java:
```diff
@@ -0,0 +1,124 @@
+    }
+    DownloadServerLogs downloadServerLogsTask = new DownloadServerLogs();
+    try {
+      return downloadServerLogsTask.execute().get();
```
again, this is blocking, so there is no reason to put the code in another thread.

aquatiko replied:
will address in a seperate PR

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/ServerHandler.java:
```diff
@@ -0,0 +1,124 @@
+  }
+
+  public void sendKeySimulationSignal() {
+    hostSyncTimestamp = System.currentTimeMillis();
```
leave a todo for a more robust time sync measurement (we need at least a few round trips averaged)

aquatiko replied:
done

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/ServerHandler.java:
```diff
@@ -0,0 +1,124 @@
+        String message = inputReader.readLine();
+        serverStartTimestamp = Long.parseLong(message);
+
+        while (message != null) {
```
does this mean the connection is closed after the last timestamp?  How do you send the results back to the server?

aquatiko replied:
@mwarton No, the connection is not closed until explicitly done so, or until the `serverHandler` object goes out of scope.
Here it only means that until the inputReader is able to get messages, store that message and read one line at a time from the socket.

mwarton replied:
I guess I'm not clear on why it would return null instead of blocking for more data if the connection is still open.  I assumed it was blocking until the connection was closed by the other side.

Seems odd, but ok for now.

aquatiko replied:
Note that the timestamps data is sent from server in a continous stream and this inputReader tries to read it line by line i.e delimiting on `\n` and at the end of that continous stream if would read a null giving us cue that the timestamps are received completely.
Also, I tested with the setup, worked as expected.

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/VideoProcessor.java:
```diff
@@ -0,0 +1,119 @@
+                      Log.d(
+                          ContentValues.TAG,
+                          "Text detected at index:" + finalI + " " + visionText.getText());
+                      setNextFrameTimeStamp();
```
This is a major concurrency problem.  You are calling this function which assumes it is called on frames in sequential order, however you are calling it from an async completion function which can be run for the frames out of order. In addition the accesses to member variables would need to be locked by mutex to prevent corruption.  You need to preserve the knowledge of which frame is doing this.

Also, is there any reason to do this here?  It could easily be done in the non-async part of the code right?

aquatiko replied:
Shifted to non-async part

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/VideoProcessor.java:
```diff
@@ -0,0 +1,119 @@
+  /** Convenience method used to assign frame timestamps to individual frames based on video
+   * recording start time and the fps of video. */
+  private void setNextFrameTimeStamp() {
+    videoFrameTimestamp.add(recordStartTime + (framesProcessed * frameDuration));
```
This is accessing 4 member variables, many of which seem to only be member variables to enable this function.  Please either pass the variables in/out, or get rid of this helper.

aquatiko replied:
Moved the member variables to function scope.

mwarton reviewed (commented):
Maybe leave the threading stuff as is and we will do another pull request.  I think we probably need to move to something more robust than AsyncTask, like CompleteableFuture.

aquatiko committed:
Simplify Lag Calculator method
